typing/typecore.ml

```diff
@@ -998,6 +997,12 @@ let check_scope_escape loc env level ty =
   with Unify trace ->
     raise(Error(loc, env, Pattern_type_clash(trace, None)))
 
+let rec split_or_until f pat =
```

Comment:
This function traverses the local groups of or-patterns (one or-pattern and also recursively its or-pattern children), calling its function argument `f` on each non-pattern children to find the first one that does not fail with a type error (see the `Error _ -> ..` exception handler on line 1400 before this patch, in the `Need_backtrack` handler of Ppat_or).

Reply:
Naming: the `_until` part is weird (until what?). Personally I would have liked to just compute a list of non-or children in-order, and then call `List.find_map` to find the first children that type-checks. So `collect_alternatives` would be a clear name, I think.

Reply:
So: we tried to go this route and in fact the code is uglier: `List.find_map` doesn't quite work as you want to re-raise the last exception in the all-failing case. But @Octachron rebased with a better-named function `find_valid_alternative`, and I'm happy with it.